### PR TITLE
Rpr support

### PR DESCRIPTION
**Pull request by Junie-774**
Initial support for RPR in the shell, needs testing but no known issues so far

### CODE REVIEW

File: src/Controller/Common.ts [hunk 1]
```diff
@@ -34,6 +35,7 @@ export const enum ShellJob {
 // can't get this automatically from a const enum
 // jobs are in the order they appear in the job guide
 export const ALL_JOBS = [
+	ShellJob.RPR,
```
Comment (AkairyuGestalter) [inline src/Controller/Common.ts:38]: We're intending to have jobs in this list in the order they appear in the job guide/in-game UI, so the RPR listing here should follow SAM

Comment (Junie-774) [reply src/Controller/Common.ts]: oop, yeah i'm dum. fixed

File: src/Game/Jobs/RPR.ts [hunk 2]
```diff
@@ -0,0 +1,1141 @@
...
+            }
+        })
+    }
+
+    setTimedResource(rscType: RPRResourceType, amount: number) {
+        const duration = (getResourceInfo(ShellJob.RPR, rscType) as ResourceInfo).maxTimeout;
+        const resource = this.resources.get(rscType);
+        resource.consume(resource.availableAmount());
+        resource.gain(amount);
+        this.enqueueResourceDrop(rscType, duration);
```
Comment (zqsz-xiv) [inline src/Game/Jobs/RPR.ts:95]:
```suggestion
        this.tryConsumeResource(rscType);
        this.resources.get(rscType).gain(amount);
        this.enqueueResourceDrop(rscType);
```

`enqueueResourceDrop` automatically retrieves `maxTimeout` of the corresponding resource, and the `tryConsumeResource` helper automatically consumes all stacks and cancels existing timers

File: src/Game/Skills.ts [hunk 3]
```diff
@@ -49,6 +49,16 @@ export function combineEffects<T extends PlayerState>(f1: EffectFn<T>, ...fs: Ar
...
 
+export function combinePredicatesAnd<T extends PlayerState>(f1: StatePredicate<T>, ...fs: Array<StatePredicate<T>>): StatePredicate<T> {
+	return (state: T) => {
+		let value: boolean = f1(state);
+		for (const fn of fs) {
+			value = (value && fn(state));
+		}
+		return value;
+	}
+}
```
Comment (zqsz-xiv) [inline src/Game/Skills.ts:60]:
```suggestion
	return (state: T) => f1(state) && fs.map((pred) => pred(state)).every();
}
```
more concise (unless you want to short-circuit in the loop)

File: src/Game/Traits.ts [hunk 4]
```diff
@@ -144,6 +156,21 @@ export class Traits {
...
+			[TraitName.ShroudGauge, 80],
+			[TraitName.EnhancedArcaneCrest, 84],
+			[TraitName.MeleeMasteryII, 84],
+			[TraitName.VoidSoul, 86],
+			[TraitName.EnhancedArcaneCircle, 88],
+			[TraitName.EnhancedEnshroud, 92],
+			[TraitName.EnhancedSecondWind, 94],
+			[TraitName.MeleeMasteryIII, 94],
+			[TraitName.EnhancedGluttony, 96],
+			[TraitName.EnhancedFeint, 98],
```
Comment (zqsz-xiv) [inline src/Game/Traits.ts:171]:
```suggestion
```

already specified for SAM (we don't have a great way of dealing with role traits right now)

File: src/Components/Jobs/RPR.tsx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                kind: "text",
+                name: "combo",
+                text: resources.get(ResourceType.RPRCombo).availableAmount().toString(),
+            },
+            {
+                kind: "text",
+                name: localize({
+                    en: "Death's Design",
+                }),
+                text: dd.toFixed(3),
```
[the remark this answers is not in the document]
Comment (Junie-774) [reply src/Components/Jobs/RPR.tsx:132]: Yeah, I made that part before I got the debuff up, this should be removed

File: src/Game/Jobs/RPR.ts [hunk 6]
```diff
@@ -0,0 +1,1141 @@
...
+makeRPRResource(ResourceType.Bloodbath, 1, {timeout: 20});
+
+export class RPRState extends GameState {
+    constructor(config: GameConfig) {
+        super(config);
+
+        const soulSliceStacks = Traits.hasUnlocked(TraitName.TemperedSoul, config.level) ? 2 : 1;
+        this.cooldowns.set(new CoolDown(ResourceType.cd_SoulSlice, 30, soulSliceStacks, soulSliceStacks));
+
+        this.cooldowns.set(new CoolDown(ResourceType.cd_ArcaneCircle, 120, 1, 1));
```
[the remark this answers is not in the document]
Comment (Junie-774) [reply src/Game/Jobs/RPR.ts:65]: tbh i'm not entirely sure what the model is supposed to be for setting cooldowns. You're probably right!

File: src/Components/Jobs/RPR.tsx [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        const shroudGauge = resources.get(ResourceType.Shroud).availableAmount();
+        const lemureShroud = resources.get(ResourceType.LemureShroud).availableAmount();
+        const voidShroud = resources.get(ResourceType.VoidShroud).availableAmount();
+        const dd = resources.timeTillReady(ResourceType.DeathsDesign);
+
+        const infos: ResourceDisplayProps[] = [
+            {
+                kind: "text",
+                name: "combo",
+                text: resources.get(ResourceType.RPRCombo).availableAmount().toString(),
```
[the remark this answers is not in the document]
Comment (Junie-774) [reply src/Components/Jobs/RPR.tsx:125]: Initially it was a counter for debugging, yeah, oops. Better to turn it into a combo timer

File: src/Game/Jobs/RPR.ts [hunk 8]
```diff
@@ -0,0 +1,1141 @@
...
+}
+
+const makeRPRSpell = (name: RPRSkillName, unlockLevel: number, params: {
+    replaceIf: ConditionalSkillReplace<RPRState>[],
+    startOnHotbar?: boolean,
+    potency: number | Array<[TraitName, number]>,
+    secondaryCooldown?: CooldownGroupProperies,
+    aspect: Aspect,
+    castTime: number | ResourceCalculationFn<RPRState>,
+    recastTime: number,
```
[the remark this answers is not in the document]
Comment (Junie-774) [reply src/Game/Jobs/RPR.ts:344]: Yeah, reaper's spells do scale with sps!

File: src/Game/Skills.ts [hunk 9]
```diff
@@ -49,6 +49,10 @@ export function combineEffects<T extends PlayerState>(f1: EffectFn<T>, ...fs: Ar
 	};
 }
 
+export function combinePredicatesAnd<T extends PlayerState>(f1: StatePredicate<T>, ...fs: Array<StatePredicate<T>>): StatePredicate<T> {
+	return (state: T) => f1(state) && fs.map((pred) => pred(state)).every(p => p);
```
Comment (zqsz-xiv) [inline src/Game/Skills.ts:53]:
```suggestion
	return (state: T) => f1(state) && fs.every((pred) => pred(state));
```

I mixed this up with the analogous python function in my original suggestion, this should be better

File: src/Components/Jobs/RPR.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (zqsz-xiv) [reply src/Components/Jobs/RPR.tsx]: seems like it only shows the single-target 1/2/3 right now, do you want it to display the AoE combo timer as well?

**Review by zqsz-xiv — APPROVED** [review]
Just a couple minor things, otherwise LGTM. Remember also to add RPR to `TESTING_JOBS` in `Controller/Common.ts` until we get more community testing.

**Comment by Junie-774** [comment]
Yeah, good catch. Tracks whichever of single-target or aoe combo is active, now!

**Pull request merged**